zebra-chain/src/serialization/compact_size.rs

```diff
@@ -0,0 +1,350 @@
+///     Cursor::new(b"\xfe\x00\x00\x20\x00").zcash_deserialize_into().unwrap(),
+/// );
+///
+/// assert_eq!(
+///     CompactSize64::from(max_size + 1),
+///     Cursor::new(b"\xfe\x01\x00\x20\x00").zcash_deserialize_into().unwrap(),
+/// );
+///
+/// assert_eq!(
+///     CompactSize64::from(0xbbaafd),
+///     Cursor::new(b"\xfe\xfd\xaa\xbb\x00").zcash_deserialize_into().unwrap(),
+/// );
+///
+/// assert_eq!(
+///     CompactSize64::from(0x22ccbbaafd),
+///     Cursor::new(b"\xff\xfd\xaa\xbb\xcc\x22\x00\x00\x00").zcash_deserialize_into().unwrap(),
+/// );
+///```
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct CompactSize64(
+    /// A numeric value.
+    ///
+    /// This field is private for consistency with [`CompactSizeMessage`].
+    u64,
+);
+
+// We don't implement conversions between CompactSizeMessage and CompactSize64,
+// because we want to distinguish fields with different numeric constraints.
+//
+// We don't implement From<CompactSizeMessage> for u16 or u8,
+// because we want all values to go through the same code paths.
+// (And we don't want to accidentally truncate using `as`.)
+
+impl TryFrom<u64> for CompactSizeMessage {
+    type Error = SerializationError;
+
+    #[inline]
+    fn try_from(size: u64) -> Result<Self, Self::Error> {
+        use SerializationError::Parse;
+
+        let size: u32 = size.try_into()?;
+
+        // # Security
+        // Defence-in-depth for memory DoS via preallocation.
+        if size
+            > MAX_PROTOCOL_MESSAGE_LEN
+                .try_into()
+                .expect("MAX_PROTOCOL_MESSAGE_LEN fits in u32")
+        {
+            // This could be invalid data from peers, so we return a parse error.
+            Err(Parse("compactsize larger than protocol message limit"))?;
+        }
+
+        Ok(CompactSizeMessage(size))
+    }
+}
+
+impl From<CompactSizeMessage> for u64 {
+    #[inline]
+    fn from(size: CompactSizeMessage) -> Self {
+        size.0.into()
+    }
+}
+
+impl From<CompactSize64> for u64 {
+    #[inline]
+    fn from(size: CompactSize64) -> Self {
+        size.0
+    }
+}
+
+impl From<u64> for CompactSize64 {
+    #[inline]
+    fn from(size: u64) -> Self {
+        CompactSize64(size)
+    }
+}
+
+impl ZcashSerialize for CompactSizeMessage {
+    /// Serialize a CompactSizeMessage into the Zcash consensus-critical format.
+    ///
+    /// # Panics
+    ///
+    /// If the value exceeds `MAX_PROTOCOL_MESSAGE_LEN`.
+    #[inline]
+    fn zcash_serialize<W: std::io::Write>(&self, writer: W) -> Result<(), std::io::Error> {
+        // # Security
+        // Defence-in-depth for memory DoS via preallocation.
+        if self.0
+            > MAX_PROTOCOL_MESSAGE_LEN
+                .try_into()
+                .expect("usize fits in u64")
+        {
+            // This is validated data inside Zebra, so we panic.
+            panic!("compactsize larger than protocol message limit");
+        }
```

Comment:
Nit: Should this be an assertion?

```suggestion
        // This is validated data inside Zebra, so we panic.
        assert!(
            self.0
                <= MAX_PROTOCOL_MESSAGE_LEN
                    .try_into()
                    .expect("usize fits in u64"),
            "compactsize larger than protocol message limit"
        );
```

Reply:
Fixed in 27cd62bf4